src/machine.adoc

```diff
@@ -2508,14 +2508,16 @@ available only to machine-mode.
 
 ==== Main Memory versus I/O versus Vacant Regions
 
-The most important characterization of a given memory address range is
-whether it holds regular main memory, or I/O devices, or is vacant.
+The most important characterization of a specific memory address range is
+whether it holds regular main memory or I/O devices, or if it is vacant.
```

Comment:
Are Oxford commas outside the consistent style for this document?

Reply:
it is really 2 ideas though - it can hold regular main memory or I/O devices.  Or it can be vacant. The , or is just to separate the 2 ideas.

Reply:
I want to be sure of the formal model.

- {Main memory, I/O, Vacant} all addresses have one and only one of these three attributes.
- {Main memory, I/O} as one exclusive set attribute and Vacant as a separate boolean attribute.  If so, are there any constraints between them?  The prose talks about I/O + Vacant, but is Main memory + Vacant also allowed?

Reply:
The distinction may or may not matter in observable behavior from the hart, but can matter for any future formalizing of the attributes, for, say, a system architecture specification to describe or control PMAs.

Reply:
It seems like what gets us into trouble is that the current PMA text views there being only two type of regions (main memory and I/O), and views a 'vacant' region as being a special type of I/O region.

Either we stick with that (which goes against some of the proposed text changes in this PR), or we switch the text to a view that there are three types of regions - in which case a 'vacant' region is no longer a special type of I/O region.

The latter seems much cleaner, but I don't know if there is some relevant history behind why a vacant region needs to be described as a type of I/O region.  @aswaterman What say you?

Reply:
The relevant original text is a little out of date with other changes to spec.  I think it will actually be cleaner to describe vacant regions as a type of IO region, and that there are really only two types of memory range (main memory and IO), so regions are either main memory or I/O, and only I/O can have vacant regions.  I think this hierarchy makes more sense from a platform perspective, where "vacancy" is really an absence of PMAs that can only exist in non-main-memory regions, with accesses to vacant regions returning access faults or bus-error interrupts.  Also, while the text recommends PMA checks be done synchronously where possible, it is acknowledged in real platforms that some checks can only happen later in an instruction's life cycle but non-synchronous exceptions should only be for IO regions not main memory.  One other aspect is that FENCE operations are only defined to distinguish memory and IO, and while vacant memory accesses do not cause an access, we might want to be able to order the effect of a bus-error interrupt wrt other accesses, so should use IO ordering for accesses to vacant regions (e.g., when probing for existence of devices).

Reply:
I support the notion of vacant regions being permissionless I/O regions. We already need to tolerate I/O regions having unusual characteristics, including re: permissions. So we can encode the vacancy property without needing any new mechanisms or characterizations.